clippy_lints/src/attrs.rs

```diff
@@ -496,35 +552,82 @@ fn is_word(nmi: &NestedMetaItem, expected: Symbol) -> bool {
+        check_deprecated_cfg_attr(cx, attr);
+        check_mismatched_target_os(cx, attr);
+    }
+}
+
+fn check_deprecated_cfg_attr(cx: &EarlyContext<'_>, attr: &Attribute) {
+    if_chain! {
+        // check cfg_attr
+        if attr.check_name(sym!(cfg_attr));
+        if let Some(items) = attr.meta_item_list();
+        if items.len() == 2;
+        // check for `rustfmt`
+        if let Some(feature_item) = items[0].meta_item();
+        if feature_item.check_name(sym!(rustfmt));
+        // check for `rustfmt_skip` and `rustfmt::skip`
+        if let Some(skip_item) = &items[1].meta_item();
+        if skip_item.check_name(sym!(rustfmt_skip)) ||
+            skip_item.path.segments.last().expect("empty path in attribute").ident.name == sym!(skip);
+        // Only lint outer attributes, because custom inner attributes are unstable
+        // Tracking issue: https://github.com/rust-lang/rust/issues/54726
+        if let AttrStyle::Outer = attr.style;
+        then {
+            span_lint_and_sugg(
+                cx,
+                DEPRECATED_CFG_ATTR,
+                attr.span,
+                "`cfg_attr` is deprecated for rustfmt and got replaced by tool attributes",
+                "use",
+                "#[rustfmt::skip]".to_string(),
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
+fn check_mismatched_target_os(cx: &EarlyContext<'_>, attr: &Attribute) {
+    fn find_mismatched_target_os(items: &[NestedMetaItem]) -> Vec<(&str, Span)> {
+        let mut mismatched = Vec::new();
+        for item in items {
+            if let NestedMetaItem::MetaItem(meta) = item {
+                match &meta.kind {
+                    MetaItemKind::List(list) => {
+                        mismatched.extend(find_mismatched_target_os(&list));
+                    },
+                    MetaItemKind::Word => {
+                        if let Some(ident) = meta.ident() {
+                            let name = &*ident.name.as_str();
+                            if let Some(os) = OPERATING_SYSTEMS.iter().find(|&&os| os == name) {
+                                mismatched.push((os, ident.span));
+                            }
+                        }
+                    },
+                    _ => {},
+                }
+            }
+        }
+        mismatched
+    }
+
+    if_chain! {
+        if attr.check_name(sym!(cfg));
+        if let Some(list) = attr.meta_item_list();
+        then {
+            let mismatched = find_mismatched_target_os(&list);
+            for (os, span) in mismatched {
+                let mess = format!("`{}` is not a valid target family", os);
+                let sugg = format!("target_os = \"{}\"", os);
+
+                span_lint_and_then(cx, MISMATCHED_TARGET_OS, span, &mess, |diag| {
+                    diag.span_suggestion(span, "try", sugg, Applicability::MaybeIncorrect);
+                    diag.help("Did you mean `unix`?");
+                });
```

Comment:
Instead of displaying a new message for each name, you could display the message once and the display multiple suggestions per `cfg`, by pulling down the for loop in the closure.

Does the `unix` suggestion apply to every OS above?

Reply:
Hmm good point, I missed `cloudabi, hermit, wasi, none` not being `unix` according to the definitions in `librustc_target/spec`.

Also, some of the OSes are missing in the test, will fix that too!